src/github.com/matrix-org/go-neb/services/slackapi/message.go

```diff
@@ -0,0 +1,177 @@
+	"mime"
+	"net/http"
+	"regexp"
+	"text/template"
+	"time"
+)
+
+type slackAttachment struct {
+	Fallback string  `json:"fallback"`
+	Color    *string `json:"color"`
+	Pretext  string  `json:"pretext"`
+
+	AuthorName *string `json:"author_name"`
+	AuthorLink *string `json:"author_link"`
+	AuthorIcon *string `json:"author_icon"`
+
+	Title     *string `json:"title"`
+	TitleLink *string `json:"title_link"`
+
+	Text     string   `json:"text"`
+	MrkdwnIn []string `json:"mrkdwn_in"`
+	Ts       *int64   `json:"ts"`
+}
+
+type slackMessage struct {
+	Text        string            `json:"text"`
+	Username    string            `json:"username"`
+	Channel     string            `json:"channel"`
+	Mrkdwn      *bool             `json:"mrkdwn"`
+	Attachments []slackAttachment `json:"attachments"`
+}
+
+// We use text.template because any fields of any attachments could
+// be Markdown, so it's convenient to escape on a field-by field basis.
+// We do not do this yet, since it's assumed that clients also escape the content we send them.
+var htmlTemplate, _ = template.New("htmlTemplate").Parse(`
+<strong>@{{ .Username }}</strong> via <strong>#{{ .Channel }}</strong><br />
+{{ if .Text }}{{ .Text }}<br />{{ end }}
+{{ range .Attachments }}
+		{{ if .AuthorName }}
+			{{if .AuthorLink }}<a href="{{ .AuthorLink }}">{{ end }}
+				{{ if .AuthorIcon }}{{ .AuthorIcon }}{{ end }}
+				{{ .AuthorName }}
+			{{if .AuthorLink }}</a>{{ end }}
+			<br />
+		{{ end }}
+	<strong><font color="{{ .Color }}">▌</font>{{ if .TitleLink }}<a href="{{ .TitleLink}}">{{ .Title }}</a>{{ else }}{{ .Title }}{{ end }}<br /></strong>
+	{{ if .Pretext }}{{ .Pretext }}<br />{{ end }}
+	{{ if .Text }}{{ .Text }}<br />{{ end }}
+{{ end }}
+`)
+
+var netClient = &http.Client{
+	Timeout: time.Second * 10,
+}
+
+var linkRegex, _ = regexp.Compile("<([^|]+)(\\|([^>]+))?>")
+
+func getSlackMessage(req http.Request) (message slackMessage, err error) {
+	ct := req.Header.Get("Content-Type")
+	ct, _, err = mime.ParseMediaType(ct)
+
+	if ct == "application/x-www-form-urlencoded" {
+		req.ParseForm()
+		payload := req.Form.Get("payload")
+		err = json.Unmarshal([]byte(payload), &message)
+	} else if ct == "application/json" {
+		log.Info("Parsing as JSON")
+		decoder := json.NewDecoder(req.Body)
+		err = decoder.Decode(&message)
+	} else {
+		message.Text = fmt.Sprint("**Error:** unknown Content-Type `%s`", ct)
+		log.Error(message.Text)
+	}
+
+	return
+}
+
+func linkifyString(text string) string {
+	return linkRegex.ReplaceAllString(text, "<a href=\"$1\">$3</a>")
+}
+
+func getColor(color *string) string {
+	if color != nil {
+		// https://api.slack.com/docs/message-attachments defines these aliases
+		mappedColor, ok := map[string]string{
+			"good":    "green",
+			"warning": "yellow",
+			"danger":  "red",
+		}[*color]
+		if ok {
+			return mappedColor
+		}
+		return *color
+	}
+	return "black"
+}
+
+func slackMessageToHTMLMessage(message slackMessage) (html matrix.HTMLMessage, err error) {
```

Comment:
Named returns for functions as long as this are really hard to read. Please stick to plain old returns in this case.